### PR TITLE
[Feature] Add PSANet

### PR DESCRIPTION
**Pull request by justld**
添加PSANet，psamask C++外部算子
说明：自定义外部算子多卡训练时不知道为啥编译不通过（也没有日志），无奈使用单卡训练，约50个小时。psanet导出需要固定input_shape，否则psamask算子输出维度会丢失。故tipc中添加input_shape参数
存在一个BUG，导出静态图后预测结果和动态图不一致，issue：https://github.com/PaddlePaddle/Paddle/issues/42068

### CODE REVIEW

File: paddleseg/models/ops/__init__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import os
+from paddle.autograd import PyLayer
```
Comment (shiyutang) [inline paddleseg/models/ops/__init__.py]: 这两个似乎没有用上？

File: paddleseg/models/psanet.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+import paddle
```
Comment (shiyutang) [inline paddleseg/models/psanet.py]: 这一行上移带14吧

File: paddleseg/models/psanet.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                 mask_h=59,
+                 mask_w=59,
+                 normalization_factor=1.0,
+                 psa_softmax=True):
+        super().__init__()
+        assert psa_type in [
+            0, 1, 2
+        ], "Only support psa type in [0, 1, 2], but got {}.".format(psa_type)
+        assert mask_h % 2 == 1 and mask_w % 2 == 1, "The value of mask_h and mask_w in point-wise spatial attention module must be odd number, but got {} and {}.".format(
+            mask_h, mask_w)
```
Comment (shiyutang) [inline paddleseg/models/psanet.py:141]: 这个格式看上去比较奇怪，重新用pre-commit跑一下？

File: paddleseg/models/psanet.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        self.mid_channels = mid_channels
+        self.psa_type = psa_type
+        self.shrink_factor = shrink_factor
+        self.mask_h = mask_h
+        self.mask_w = mask_w
+        self.psa_softmax = psa_softmax
+        if normalization_factor is None:
+            normalization_factor = mask_h * mask_w
```
Comment (shiyutang) [inline paddleseg/models/psanet.py:149]: 这一行直接self.normalization_factor = mask_h*mask_w?

**Review by shiyutang — CHANGES_REQUESTED** [review]
细节可以完善一下

File: paddleseg/models/ops/psamask.cu [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+#include <vector>
+
+#define CHECK_GPU_INPUT(x) PD_CHECK(x.place() == paddle::PlaceType::kGPU, #x " must be a GPU Tensor.")
+
+#ifndef min
+#define min(a,b) (((a) < (b)) ? (a) : (b))
+#endif
+
+#ifndef max
+#define max(a,b) (((a) > (b)) ? (a) : (b))
```
Comment (shiyutang) [inline paddleseg/models/ops/psamask.cu:12]: @michaelowenliu 需要查看下增加op的位置和方式是否合理。

Comment (shiyutang) [reply paddleseg/models/ops/psamask.cu:12]: @juncaipeng 看下新增算子的实现方式呢？

File: paddleseg/models/psanet.py (hunk 4 above)
Comment (justld) [reply paddleseg/models/psanet.py:149]: 这里和参考代码一致：https://github.com/hszhao/semseg/blob/4f274c3f276778228bc14a4565822d46359f0cc8/model/psanet.py#L20

File: paddleseg/models/ops/__init__.py (hunk 1 above)
Comment (justld) [reply paddleseg/models/ops/__init__.py]: 已删除

File: paddleseg/models/psanet.py (hunk 3 above)
Comment (justld) [reply paddleseg/models/psanet.py:141]: 跑过了，就是这样

**Comment by justld** [comment]
训练：单卡Tesla V100 约50小时
目前导出静态图后预测与动态图不一致：https://github.com/PaddlePaddle/Paddle/issues/42068

**Comment by justld** [comment]
image
image
image
image
image

test_tipc约3分钟，（导出的静态图预测与动态图不一致，见上方issue连接）

**Comment by justld** [comment]
aistudio测试连接：https://aistudio.baidu.com/aistudio/projectdetail/3851842?contributionType=1

File: paddleseg/models/psanet.py (hunk 2 above)
Comment (justld) [reply paddleseg/models/psanet.py]: 已修改
